opentelekomcloud/utils.go

```diff
@@ -18,3 +22,36 @@ func jsonBytesEqual(b1, b2 []byte) bool {
 
 	return reflect.DeepEqual(o1, o2)
 }
+
+// convertStructToMap converts an instance of struct to a map object, and
+// changes each key of fileds to the value of 'nameMap' if the key in it
+// or to its corresponding lowercase.
+func convertStructToMap(obj interface{}, nameMap map[string]string) (map[string]interface{}, error) {
+	b, err := json.Marshal(obj)
+	if err != nil {
+		return nil, fmt.Errorf("refreshResourceData:: marshal failed:%v", err)
+	}
+
+	m := regexp.MustCompile(`"[a-z0-9A-Z_]+":`)
```

Comment:
move this outside the `func` so it can fail earlier if a problem exists

Reply:
Sorry, I am not quite understand. Do you mean move line 35 outside of method of 'convertStructToMap'? Why? Line 35 will not be panic because it passes an raw string to 'regexp.MustCompile'.

Reply:
It can panic if the string is a bad regex. See [the docs](https://golang.org/pkg/regexp/#MustCompile), the `Must*` methods are typically for global initialization:

> MustCompile is like Compile but panics if the expression cannot be parsed. It simplifies safe initialization of global variables holding compiled regular expressions.

Reply:
It may panic if regexp.MustCompile receives a variable whose value is inconclusive. But at line 35, it receives a valid regular express, so it can not panic. I can use regexp.Compile instead of it, which will not lead to panic.